File: .idea/app/src/main/java/com/chevstrap/rbx/JsonManager.java
package com.chevstrap.rbx;

import com.chevstrap.rbx.Utility.FileTool;
import com.chevstrap.rbx.Utility.MD5Hash;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.*;
        import java.util.*;

public abstract class JsonManager<T extends Map<String, Object>> {
    protected T prop;
    protected T originalProp;
    protected boolean loaded = false;
    private String lastFileHash = null;

    public abstract String getClassName();
    public abstract File getFileLocation();
    protected abstract T createEmptyMap();

    public boolean isLoaded() {
        return loaded;
    }

    public void save() {
        File file = getFileLocation();
        JSONObject json = new JSONObject(prop);
        try {
            FileTool.safeWrite(file, json.toString(4));
            lastFileHash = MD5Hash.fromFile(file.getAbsolutePath());
        } catch (IOException | JSONException e) {
            throw new RuntimeException("Failed to save " + getClassName() + ": " + e.getMessage(), e);
        }
    }

    public void load(boolean alertFailure) {
        String LOG_IDENT = "JsonManager::Load";
        File file = getFileLocation();

        if (!file.exists()) {
            prop = createEmptyMap();
            loaded = false;
            lastFileHash = null;
            return;
        }

        try {
            String currentHash = MD5Hash.fromFile(file.getAbsolutePath());

            if (!hasFileChanged() && loaded) {
                return;
            }

            if (currentHash != null && currentHash.equals(lastFileHash)) {
                loaded = true;
                return;
            }

            String jsonData = FileTool.read(file);
            JSONObject jsonObject = new JSONObject(jsonData);

            prop = convertJsonObjectToMap(jsonObject);

            lastFileHash = currentHash;
            loaded = true;

        } catch (IOException | JSONException e) {
            if (alertFailure) {
                App.getLogger().writeLine(LOG_IDENT, "Failed to load " + getClassName() + ": " + e.getMessage());
            }

            try {
                if (file.exists()) {
                    String backupName = getClassName() + "_backup.json";
                    File backupFile = new File(file.getParentFile(), backupName);
                    FileTool.copy(file, backupFile, true);

                    App.getLogger().writeLine(LOG_IDENT, "Backup created: " + backupFile.getAbsolutePath());
                }
            } catch (Exception copyEx) {
                App.getLogger().writeLine(LOG_IDENT, "Failed to create backup: " + copyEx.getMessage());
            }

            prop = createEmptyMap();
            loaded = false;
            lastFileHash = null;
            save();
        }
    }

    private T convertJsonObjectToMap(JSONObject jsonObject) throws JSONException {
        T map = createEmptyMap();
        JSONArray keys = jsonObject.names();
        if (keys == null) return map;

        for (int i = 0; i < keys.length(); i++) {
            String key = keys.getString(i);
            Object value = jsonObject.get(key);

            if (value instanceof JSONObject) {
                value = convertJsonObjectToMap((JSONObject) value);
            } else if (value instanceof JSONArray) {
                value = convertJsonArrayToList((JSONArray) value);
            }
            map.put(key, value);
        }
        return map;
    }

    private List<Object> convertJsonArrayToList(JSONArray array) throws JSONException {
        List<Object> list = new ArrayList<>();

        for (int i = 0; i < array.length(); i++) {
            Object value = array.get(i);

            if (value instanceof JSONObject) {
                value = convertJsonObjectToMap((JSONObject) value);
            } else if (value instanceof JSONArray) {
                value = convertJsonArrayToList((JSONArray) value);
            }
            list.add(value);
        }

        return list;
    }

    public String getCurrentFileHash() {
        File file = getFileLocation();
        if (file.exists()) {
            return MD5Hash.fromFile(file.getAbsolutePath());
        }
        return null;
    }

    public boolean hasFileChanged() {
        String currentHash = getCurrentFileHash();
        return currentHash != null && !currentHash.equals(lastFileHash);
    }
}

File: app/src/main/java/com/chevstrap/rbx/JsonManager.kt
package com.chevstrap.rbx

import chevstrap.extensions.FileTool
import com.chevstrap.rbx.utility.MD5Hash
import kotlinx.serialization.KSerializer
import kotlinx.serialization.json.Json
import java.io.File
import java.io.IOException

abstract class JsonManager<T> {

    @JvmField
    protected var prop: T? = null

    @JvmField
    protected var originalProp: T? = null

    var isLoaded: Boolean = false
        protected set

    private var lastFileHash: String? = null

    protected val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
        coerceInputValues = true
    }

    abstract val className: String

    abstract val fileLocation: File

    protected abstract val serializer: KSerializer<T>

    protected abstract fun createDefault(): T

    open fun save() {
        val file = fileLocation
        val value = prop ?: createDefault()

        try {
            val jsonData = json.encodeToString(
                serializer,
                value
            )

            FileTool.safeWrite(
                file,
                jsonData
            )

            lastFileHash = MD5Hash.fromFile(
                file.absolutePath
            )

            isLoaded = true
        } catch (e: IOException) {
            throw RuntimeException(
                "Failed to save $className: ${e.message}",
                e
            )
        } catch (e: Exception) {
            throw RuntimeException(
                "Failed to save $className: ${e.message}",
                e
            )
        }
    }

    open fun load(alertFailure: Boolean) {
        val file = fileLocation

        if (!file.exists()) {
            prop = createDefault()
            isLoaded = false
            lastFileHash = null
            return
        }

        try {
            val currentHash = MD5Hash.fromFile(
                file.absolutePath
            )

            if (!hasFileChanged() && isLoaded) {
                return
            }

            if (
                currentHash != null &&
                currentHash == lastFileHash
            ) {
                isLoaded = true
                return
            }

            val jsonData = FileTool.safeRead(file)

            prop = json.decodeFromString(
                serializer,
                jsonData
            )

            lastFileHash = currentHash
            isLoaded = true

        } catch (e: Exception) {
            handleLoadError(
                file,
                alertFailure,
                e
            )
        }
    }

    private fun handleLoadError(
        file: File,
        alertFailure: Boolean,
        exception: Exception
    ) {
        if (alertFailure) {
            App.logger.writeLine(
                "JsonManager::Load",
                "Failed to load $className: ${exception.message}"
            )
        }

        try {
            if (file.exists()) {
                val backupName = "${className}_backup.json"
                val backupFile = File(
                    file.parentFile,
                    backupName
                )

                FileTool.copy(
                    file,
                    backupFile,
                    true
                )

                App.logger.writeLine(
                    "JsonManager::Load",
                    "Backup created: ${backupFile.absolutePath}"
                )
            }
        } catch (copyException: Exception) {
            App.logger.writeLine(
                "JsonManager::Load",
                "Failed to create backup: ${copyException.message}"
            )
        }

        prop = createDefault()
        isLoaded = false
        lastFileHash = null

        save()
    }

    protected fun deepEquals(
        a: Any?,
        b: Any?
    ): Boolean {
        if (a === b) {
            return true
        }

        if (a == null || b == null) {
            return false
        }

        return when (a) {
            is Number if b is Number -> {
                a.toDouble() == b.toDouble()
            }

            is Map<*, *> if b is Map<*, *> -> {
                if (a.size != b.size) {
                    return false
                }

                a.all { (key, value) ->
                    b.containsKey(key) &&
                            deepEquals(
                                value,
                                b[key]
                            )
                }
            }

            is List<*> if b is List<*> -> {
                if (a.size != b.size) {
                    return false
                }

                a.indices.all { index ->
                    deepEquals(
                        a[index],
                        b[index]
                    )
                }
            }

            else -> {
                a == b
            }
        }
    }

    val currentFileHash: String?
        get() {
            val file = fileLocation

            return if (file.exists()) {
                MD5Hash.fromFile(
                    file.absolutePath
                )
            } else {
                null
            }
        }

    fun hasFileChanged(): Boolean {
        val currentHash = currentFileHash

        return currentHash != null &&
                currentHash != lastFileHash
    }
}